### DISCUSSION TITLE
persistent cache build dependency warning when "module" imported from esm build dependency

### DISCUSSION THREAD

**Discussion by youngderekm** [webpack/webpack#14645] [General]
# Bug report

**What is the current behavior?**
My webpack configuration is an es module (webpack.config.mjs).  It imports a file from another es module (type: "module") that imports `createRequire` from "module" in order to use `require.resolve`.  I have cache.type = 'filesystem'.  When the build dependencies are scanned, I get a warning that 'module' can't be resolved:

```
<w> [webpack.cache.PackFileCacheStrategy] Caching failed for pack: Error: Can't resolve 'module' in 'C:\sand\webinst-webpack\webcommons\tools\client\webpack\target\tsc'
<w> while resolving 'module' in C:\sand\webinst-webpack\webcommons\tools\client\webpack\target\tsc as file
<w>  at resolve esm file module
<w>  at file dependencies C:\sand\webinst-webpack\webcommons\tools\client\webpack\target\tsc\WebpackConfig.js
<w>  at file C:\sand\webinst-webpack\webcommons\tools\client\webpack\target\tsc\WebpackConfig.js
<w>  at resolve commonjs C:\sand\webinst-webpack\webcommons\tools\client\webpack\target\tsc\WebpackConfig.js
```
 

**If the current behavior is a bug, please provide the steps to reproduce.**

In webpack.config.mjs:

```
import {WebpackConfig} from "webpack-config";
```

from webpack-config/index.js:

```
import {createRequire} from "module";
const require = createRequire(import.meta.url);

....

    cache: {
      type: "filesystem",
    }
```
webpack-config/package.json has `"type": "module"`.

use `webpack serve`.
warning is logged to the console.

**What is the expected behavior?**
Imports from built in node modules (like 'module') should be ignored.

**Other relevant information:**
webpack version:  5.51.2
Node.js version: v16.13.0
Operating System: Windows 10
Additional tools:

**Comment by alexander-akait** [comment]
I don't think it is related to webpack (look at stack trace), if you need help please create reproducible example and I will show you how to fix

**Reply by youngderekm** [reply]
My reproducible case is here: https://github.com/youngderekm/webpack-cache-esm.

It should cause two types of warnings, the "Can't resolve 'module'" warning I reported above, and a different warning, "SyntaxError: Unexpected token ' in JSON".  I believe the JSON syntax error is because FileSystemInfo.js expects imports to be double quotes (so parseable as JSON strings), but in my example I have some that are single quoted.


FileSystem.js:1659
```
	// import ... from "..."
	dependency = JSON.parse(
		source.substring(imp.s - 1, imp.e + 1)
	);
```

If you disable the cache: "filesystem" setting, then there are no warnings logged.

**Reply by alexander-akait** [reply]
Weird, I can't reproduce... All works fine

**Reply by youngderekm** [reply]
That's strange.  I just simplified the project which I'm hoping will make it easier to reproduce.  Also, I put instructions at the end of the README that show how to run it under docker if that's helpful to get a consistent environment.  My simplified project is at the same location: https://github.com/youngderekm/webpack-cache-esm

**Reply by alexander-akait** [reply]
What is `Node.js` version? Can you check it in docker?

**Reply by youngderekm** [reply]
node 16.3.0 or16.4.0 is used.

To reproduce with docker, I used the `node:16` image, which currently has 16.3.0.  There is no Dockerfile necessary.  I just did this:

    docker run -v ${PWD}:/src -it --rm node:16 /bin/bash

or

   docker run -v \`pwd\`:/src -it --rm node:16 /bin/bash

now I'm in a shell in the docker container with the source from the repo at /src.  Do the following to reproduce:

    cd /src
    rm -rf node_modules
    npm install
    npm run build

You should see the warnings.

**Reply by alexander-akait** [reply]
Weird bug, if you change `import {createRequire} from "module";` on single `import {createRequire} from 'module';` all works fine (except we can't resolve Node.js deps, but it is expected for ESM), problems in quotes, maybe bug in `es-module-lexer`, investigating

**Reply by alexander-akait** [reply]
https://github.com/webpack/webpack/pull/14709/, sorry for delay, I moved it to discussion very early, sorry again

**Reply by alexander-akait** [reply]
Solved https://github.com/webpack/webpack/pull/14687/, update lock file

**Reply by youngderekm** [reply]
Thank you for looking into this.  I tried again with 5.63.0 that has the quote fix and now single quotes or double quotes produce the same behavior.  But I still get the warning about node modules:

```
Caching failed for pack: Error: Can't resolve 'module' in 'C:\sand\webpack-cache-esm'
```

And when that happens, nothing is written into the cache (node_modules/.cache isn't created).  Shouldn't node dependencies (path, module, url, etc.) just be ignored?  Is it possible to use the filesystem cache with an esm config file with node dependencies?

**Reply by alexander-akait** [reply]
Unfortunately, this is a limitation on Node.js side (we are working on it), honestly ESM part on Node.js side is not finished, many things are experimental, you can move them into `file.cjs` file and do logic there, and only export evaluated variables, yes, it is annoying and inconvenient, but no solution right now, hope we will improve it in near future

**Comment by weimengxi** [comment]
I've met the same problem.

**Reply by vankop** [reply]
should be solved in latest webpack version. feel free to create repro.

**Comment by yejinjian** [comment]
I have met the same problem in version 5.93.0

**Reply by alexander-akait** [reply]
Please create reproducible test repo
